## Preprocessing

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

In [2]:
#  Import and read the charity_data.csv.
# application_df = pd.read_csv("Resources/charity_data.csv")
url = "https://raw.githubusercontent.com/harsh-env/Deep_Learning_ML/main/Resources/charity_data.csv"
application_df=pd.read_csv(url)
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [3]:
# Drop the non-beneficial ID column, 'EIN'.
application_df = application_df.drop(columns = ['EIN'])

In [4]:
# Determine the number of unique values in each column.
for x in application_df.columns:
    print(x, len(application_df[x].unique()))

NAME 19568
APPLICATION_TYPE 17
AFFILIATION 6
CLASSIFICATION 71
USE_CASE 5
ORGANIZATION 4
STATUS 2
INCOME_AMT 9
SPECIAL_CONSIDERATIONS 2
ASK_AMT 8747
IS_SUCCESSFUL 2


In [5]:
application_df["NAME"].value_counts()

PARENT BOOSTER USA INC                                                  1260
TOPS CLUB INC                                                            765
UNITED STATES BOWLING CONGRESS INC                                       700
WASHINGTON STATE UNIVERSITY                                              492
AMATEUR ATHLETIC UNION OF THE UNITED STATES INC                          408
                                                                        ... 
ST LOUIS SLAM WOMENS FOOTBALL                                              1
AIESEC ALUMNI IBEROAMERICA CORP                                            1
WEALLBLEEDRED ORG INC                                                      1
AMERICAN SOCIETY FOR STANDARDS IN MEDIUMSHIP & PSYCHICAL INVESTIGATI       1
WATERHOUSE CHARITABLE TR                                                   1
Name: NAME, Length: 19568, dtype: int64

In [6]:
# Look at NAME value counts for binning
name_cnt = application_df["NAME"].value_counts()
name_cnt

# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `name_to_replace`
name_to_replace = list(name_cnt[name_cnt<5].index)

# Replace in dataframe
for name in name_to_replace:
    application_df['NAME'] = application_df['NAME'].replace(name,"Other")

# Check to make sure binning was successful
application_df['NAME'].value_counts()

Other                                                                19803
PARENT BOOSTER USA INC                                                1260
TOPS CLUB INC                                                          765
UNITED STATES BOWLING CONGRESS INC                                     700
WASHINGTON STATE UNIVERSITY                                            492
                                                                     ...  
GUILDS OF THE SANTA FE OPERA INC                                         5
FAMILY MOTOR COACH ASSOCIATION                                           5
JAMESTOWN SOCIETY                                                        5
MOST WORSHIPFUL PRINCE HALL GRAND LODGE OF FREE & ACCEPTED MASONS        5
PTA HAWAII CONGRESS                                                      5
Name: NAME, Length: 403, dtype: int64

In [7]:
# Look at APPLICATION_TYPE value counts for binning
counts = application_df['APPLICATION_TYPE'].value_counts()
counts

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [8]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
application_types_to_replace = list(counts[counts<400].index)
application_types_to_replace

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [9]:
# Look at CLASSIFICATION value counts for binning
cbinning = application_df['CLASSIFICATION'].value_counts()
cbinning

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [10]:
# You may find it helpful to look at CLASSIFICATION value counts >1
classificationcnts= cbinning[cbinning>1]
classificationcnts

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: CLASSIFICATION, dtype: int64

In [11]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
classifications_to_replace = list (cbinning[cbinning<500].index)
classifications_to_replace

# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")
    
# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
Other     1484
C7000      777
Name: CLASSIFICATION, dtype: int64

In [12]:
# Convert categorical data to numeric with `pd.get_dummies`
application_df = pd.get_dummies(application_df,dtype=float)
application_df.head()

STATUS  ASK_AMT  IS_SUCCESSFUL  NAME_AACE INTERNATIONAL  \
0       1     5000              1                      0.0   
1       1   108590              1                      0.0   
2       1     5000              0                      0.0   
3       1     6692              1                      0.0   
4       1   142590              1                      0.0   

   NAME_ACE MENTOR PROGRAM OF AMERICA INC  NAME_ACTS MINISTRY  \
0                                     0.0                 0.0   
1                                     0.0                 0.0   
2                                     0.0                 0.0   
3                                     0.0                 0.0   
4                                     0.0                 0.0   

   NAME_ACTS MISSIONS  \
0                 0.0   
1                 0.0   
2                 0.0   
3                 0.0   
4                 0.0   

   NAME_AFRICAN-AMERICAN POSTAL LEAGUE UNITED FOR SUCCESS A-PLUS  \
0                                                0.0               
1                                                0.0               
2                                                0.0               
3                                                0.0               
4                                                0.0               

   NAME_AIR FORCE ASSOCIATION  NAME_ALABAMA FEDERATION OF WOMENS CLUBS  ...  \
0                         0.0                                      0.0  ...   
1                         0.0                                      0.0  ...   
2                         0.0                                      0.0  ...   
3                         0.0                                      0.0  ...   
4                         0.0                                      0.0  ...   

   INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0                0.0                     0.0                       0.0   
1                1.0                     0.0                       0.0   
2                0.0                     0.0                       0.0   
3                0.0                     1.0                       0.0   
4                0.0                     0.0                       1.0   

   INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                 0.0               0.0                     0.0   
1                 0.0               0.0                     0.0   
2                 0.0               0.0                     0.0   
3                 0.0               0.0                     0.0   
4                 0.0               0.0                     0.0   

   INCOME_AMT_50M+  INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  \
0              0.0                0.0                       1.0   
1              0.0                0.0                       1.0   
2              0.0                0.0                       1.0   
3              0.0                0.0                       1.0   
4              0.0                0.0                       1.0   

   SPECIAL_CONSIDERATIONS_Y  
0                       0.0  
1                       0.0  
2                       0.0  
3                       0.0  
4                       0.0  

[5 rows x 448 columns]

In [13]:
# Split our preprocessed data into our features and target arrays
y = application_df['IS_SUCCESSFUL'].values
X = application_df.drop('IS_SUCCESSFUL', axis=1).values

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X,y,random_state = 42)

In [14]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)
len(X_train_scaled[0])

447

## Compile, Train and Evaluate the Model

In [15]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
num_features = len(X_train_scaled[0])
hidden_node_1 = 160
hidden_node_2 = 30
hidden_node_3= 18

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_node_1, input_dim=num_features, activation='relu'))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_node_2, activation='relu'))

# Third layer
nn.add(tf.keras.layers.Dense(units=hidden_node_3, activation='relu'))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 160)               71680     
                                                                 
 dense_1 (Dense)             (None, 30)                4830      
                                                                 
 dense_2 (Dense)             (None, 18)                558       
                                                                 
 dense_3 (Dense)             (None, 1)                 19        
                                                                 
Total params: 77,087
Trainable params: 77,087
Non-trainable params: 0
_________________________________________________________________


In [16]:
# Compile the model
nn.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [17]:
# Train the model
fit_model = nn.fit(X_train_scaled,y_train,epochs=60)

Epoch 1/60
804/804 [==============================] - 6s 5ms/step - loss: 0.4620 - accuracy: 0.7801
Epoch 2/60
804/804 [==============================] - 2s 2ms/step - loss: 0.4270 - accuracy: 0.7957
Epoch 3/60
804/804 [==============================] - 2s 3ms/step - loss: 0.4229 - accuracy: 0.7985
Epoch 4/60
804/804 [==============================] - 3s 3ms/step - loss: 0.4199 - accuracy: 0.7994
Epoch 5/60
804/804 [==============================] - 3s 3ms/step - loss: 0.4195 - accuracy: 0.7987
Epoch 6/60
804/804 [==============================] - 3s 3ms/step - loss: 0.4173 - accuracy: 0.8004
Epoch 7/60
804/804 [==============================] - 2s 2ms/step - loss: 0.4170 - accuracy: 0.8004
Epoch 8/60
804/804 [==============================] - 2s 3ms/step - loss: 0.4160 - accuracy: 0.8012
Epoch 9/60
804/804 [==============================] - 2s 3ms/step - loss: 0.4150 - accuracy: 0.8034
Epoch 10/60
804/804 [==============================] - 3s 3ms/step - loss: 0.4142 - accuracy: 0.8020

In [18]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.4936 - accuracy: 0.7893 - 511ms/epoch - 2ms/step
Loss: 0.4936482310295105, Accuracy: 0.7892711162567139


In [19]:
# Export our model to HDF5 file
from google.colab import files

nn.save('/content/AlphabetSoupCharity_Optimization1.h5')
files.download('/content/AlphabetSoupCharity_Optimization1.h5')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## Step 4: Neural Network Final Report

### Overview
The nonprofit foundation, Alphabet Soup needed a tool that can help it select the applicants for funding with the best chance of success in their ventures. Using machine learning and neural networks, an algorithm is developed to predict the applicant's success in their venture using various features. This algorithm will enable the foundation to assist in deciding whether to select the applicant for funding.

### Results:
##### Data Processing
In the preprocessing of data, IS_SUCCESSFUL variable was identified as the target. The purpose of the model is to determine the success of the applicant which is dependent on how effectively they use the fund. 

#### AlphabetSoupCharity_Optimization1.ipynb: 
1. Only the EIN column was dropped, & KEPT THE NAME Column to see if it increases model accuracy. The remaining variables in the dataset were retained as features. Binning were done to reduce the noise on our model.
2. IS_SUCCESSFUL variable was identified as the target
3. OPTIMIZATION TRIAL1 Model run: 160, 30 & 18 neurons for 1ST, 2ND & 3RD layers were maintained from previous run. 
4. Relu was used in 1ST, 2ND & 3RD activation and single neuron in 4TH layer with sigmoid activation
5. Epochs were reduced to 60

#### Accuracy Did Not change with increasing neurons and Epochs - 78.9%

### Summary: 
The initial model, Starter_Code.ipynb produced an accuracy below 75% which was not satisfactory, where columns EIN & NAME were dropped as they were not thought of features or target. In AlphabetSoupCharity_Optimization.ipynb, another layer was added and neurons were doubled in the 1st layer to see if it affected the model accuracy. As a result, I analyzed the dataset and realized that adding back the NAME as a feature is relevant as this determines the number of times the applicant applied for a funding. However, there is too much noise (variance) on NAME, which had to be reduced through binning.

In order to optimize the model, AlphabetSoupCharity_Optimization1.ipynb, I needed to reduce the noise caused by the NAME variable by further binning names with count of less than 5 (reducing the uniques from 19568 to 403). The third layer was maintained from earlier run which had 160, 30 and 18 neurons, for the 1st, 2nd and 3rd layers. The epoch was also reduced to 60 from 100 from previous run. The output layer's activation function was retained to sigmoid as the model is supposed classify whether the applications will be successful or not. These changes enable the model to produce an accuracy rate of above 75% or 79%.